extensions/ql-vscode/src/logging.ts

```diff
@@ -1,42 +1,124 @@
-import { window as Window, OutputChannel, Progress } from 'vscode';
+import { window as Window, OutputChannel, Progress, ExtensionContext, Disposable } from 'vscode';
 import { DisposableObject } from 'semmle-vscode-utils';
+import * as fs from 'fs-extra';
+import * as path from 'path';
+
+interface LogOptions {
+  /** If false, don't output a trailing newline for the log entry. Default true. */
+  trailingNewline?: boolean;
+
+  /** If specified, add this log entry to the log file at the specified location. */
+  additionalLogLocation?: string;
+}
 
 export interface Logger {
-  /** Writes the given log message, followed by a newline. */
-  log(message: string): void;
-  /** Writes the given log message, not followed by a newline. */
-  logWithoutTrailingNewline(message: string): void;
+  /** Writes the given log message, optionally followed by a newline. */
+  log(message: string, options?: LogOptions): Promise<void>;
   /**
    * Reveal this channel in the UI.
    *
    * @param preserveFocus When `true` the channel will not take focus.
    */
   show(preserveFocus?: boolean): void;
+
+  /**
+   * Remove the log at the specified location
+   * @param location log to remove
+   */
+  removeAdditionalLogLocation(location: string): Promise<void>;
 }
 
 export type ProgressReporter = Progress<{ message: string }>;
 
 /** A logger that writes messages to an output channel in the Output tab. */
 export class OutputChannelLogger extends DisposableObject implements Logger {
   public readonly outputChannel: OutputChannel;
+  private readonly additionalLocations = new Map<string, AdditionalLogLocation>();
+  private additionalLogLocationPath: string | undefined;
 
-  constructor(title: string) {
+  constructor(private title: string) {
     super();
     this.outputChannel = Window.createOutputChannel(title);
     this.push(this.outputChannel);
   }
 
-  log(message: string) {
-    this.outputChannel.appendLine(message);
+  init(ctx: ExtensionContext): void {
+    this.additionalLogLocationPath = path.join(ctx.storagePath || ctx.globalStoragePath, this.title);
+
+    // clear out any old state from previous runs
+    fs.remove(this.additionalLogLocationPath);
   }
 
-  logWithoutTrailingNewline(message: string) {
-    this.outputChannel.append(message);
+  /**
+   * This function is asynchronous and will only resolve once the message is written
+   * to the side log (if required). It is not necessary to await the results of this
+   * function if you don't need to guarantee that the log writing is complete before
+   * continuing.
+   */
+  async log(message: string, options = { } as LogOptions): Promise<void> {
```

Comment:
This idiom makes sense to me, but I think we may have a LGTM-based CI hook that complains about calls to async functions in void context that don't await... so we might want a wrapper function `logAsync` or some such around doing exactly that which includes whatever magic comment makes LGTM ignore it. Not a request for a change, just a heads-up.

Reply:
(at least I think that was the mechanism that caught an actual missing-`await` bug of mine a while ago...)

Reply:
Doesn't seem to be causing any errors right now. There is a eslint config for this. We can enable it. Might be prudent...